initial implementation of scale type

orndorffgrant opened the pull request:
If this PR fixes a bug, you _must_ add test cases representative of the bug.

Please refer to [Tettra](https://app.tettra.co/teams/amida/pages/amida-pull-request-and-code-review-guide) for PR review guidelines.

#### What does this PR do?
This adds the "scale" question type, which is allows min and max limits on a number input. This was adapted from a similar feature added to survey service. See https://github.com/amida-tech/amida-survey-microservice/pull/38

#### Related JIRA tickets:
RR-756

#### How should this be manually tested?
Create/get/export/import scale questions and answer them.

1. Scale questions are specified using type scale and property meta: { scaleLimits: {min: , max: }}.
2. Min and max are optional but at least one of them should be present.
3. When both present max needs to be greater or equal to min.
4. Answers to scale questions are specified using numberValue which accepts float.
5. Answers will be validated when posted according to min and max.

#### Background/Context

#### Screenshots (if appropriate):

orndorffgrant commented:
Pretty sure the swagger.json additions are unnecessary/wrong but wanted to get this up for review.

orndorffgrant commented:
This is currently its own question type and accepts float values. That was the feature added to survey service so I just did the same thing.

Now that I understand the code base better I could change it to just be extra validations on integer or float questions iff there are limits in the meta field.

unprofessional commented:
There are a couple of eslint errors:
```
/lib/json-schema.js 607:13  error  Duplicate key 'meta'  no-dupe-keys
/test/util/generator/answerer.js97:10  error  Expected 'this' to be used by class method 'scale'  class-methods-use-this
```
Use `grunt eslint` to see them.

unprofessional commented:
Also one test is failing for me:
``
(migration.spec.js 137)
1) migration spec compare question_type records:

      AssertionError: expected [ Array(17) ] to deeply equal [ Array(18) ]
      + expected - actual

         {
           "name": "pounds"
         }
         {
      +    "name": "scale"
      +  }
      +  {
           "name": "text"
         }
         {
           "name": "year"
```
It looks like the `question_type` migration-model disagrees with the canonical model. This could be be because I neglected to do something before running the tests, but these tests are self-contained (i.e. this one runs a separate `recregmigrations` database instance).

@au2 , would Grant need to perform a migration for the updated value-in-array `scale` in the canonical `question_type` model? Or am I way off base here?

au2 commented:
yes, we need a migration for the added type and columns

On Mon, Feb 5, 2018 at 11:37 AM, Unprofessional <notifications@github.com>
wrote:

> Also one test is failing for me:
> ``
> (migration.spec.js 137)
>
>    1.
>
>    migration spec compare question_type records:
>
>    AssertionError: expected [ Array(17) ] to deeply equal [ Array(18) ]
>    -
>
>       expected - actual
>
>       {
>       "name": "pounds"
>       }
>       {
>       -
>
>       "name": "scale"
>       -
>
>       }
>       -
>
>       {
>       "name": "text"
>       }
>       {
>       "name": "year"
>
> It looks like the `question_type` migration-model disagrees with the canonical model. This could be be because I neglected to do something before running the tests, but these tests are self-contained (i.e. this one runs a separate `recregmigrations` database instance).
>
> @au2 , would Grant need to perform a migration for the updated value-in-array `scale` in the canonical `question_type` model? Or am I way off base here?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/amida-tech/recruitment-registry/pull/249#issuecomment-363141182>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF1-EXubXTsqOWm3gfdOD23KRogP4zW6ks5tRy4sgaJpZM4R2F_Z>
> .
>

orndorffgrant commented:
I removed the scaleLimits defs from swagger.json because they are being sent inside the meta field.

I don't fully understand the migration tests. The change I made (adding 'scale' to the list of question types) lets the test pass but do I need to do anything else?

au2 commented:
If there are existing customers anything that changes the database needs a migration script that can be applied when customers update. Unlike services this project has an existing customer. We use sequelize-cli for migration. Take a look at the migrations directory to see some up the previous migration scripts.

au2 commented:
I do not think I used meta for scale questions in survey service. So I am not sure what you are referring to. I will try to take a look at the code today sometime.

orndorffgrant commented:
Thank you for the explanation @au2 !

I changed the implementation - partially because Alan said something about using the meta field, largely because I knew I'd learn more if I did something beyond copy/pasting.

Which means this pull request is not the exactly the same even though it serves the same purpose (and is why I called the swagger.json additions unnecessary).

In this implementation, the client includes `scaleLimits { ... }` inside `meta` in it's request. Then the backend stores `scaleLimits` inside `meta` in the db - so there is not a new `parameter` column.

Though, if we want this backend's api to be the same as survey-service then I should change things back to how you did them.

au2 commented:
Fair. But if you are going to use Meta you do not need the new question
type either. You can just tag that along to meta as well with a number or
integer question type. Either way.

On Mon, Feb 5, 2018 at 2:24 PM, Grant Orndorff <notifications@github.com>
wrote:

> Thank you for the explanation @au2 <https://github.com/au2> !
>
> I changed the implementation - partially because Alan said something about
> using the meta field, largely because I knew I'd learn more if I did
> something beyond copy/pasting.
>
> Which means this pull request is not the exactly the same even though it
> serves the same purpose (and is why I called the swagger.json additions
> unnecessary).
>
> In this implementation, the client includes scaleLimits { ... } inside
> meta in it's request. Then the backend stores scaleLimits inside meta in
> the db - so there is not a new parameter column.
>
> Though, if we want this backend's api to be the same as survey-service
> then I should change things back to how you did them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/amida-tech/recruitment-registry/pull/249#issuecomment-363193322>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF1-EciVK1Vil4g2fQ5SO80V8GPzXwCKks5tR1WEgaJpZM4R2F_Z>
> .
>

orndorffgrant commented:
To keep future migration to use survey service  easier, I changed it back to be the same as https://github.com/amida-tech/amida-survey-microservice/pull/38.

I added a migration that just adds the parameter column to the question table with default value null.

au2 commented on migration/models/question-type.model.js:
```diff
@@ -24,7 +24,7 @@ module.exports = function Table(sequelize, DataTypes) {
                         'text', 'choice', 'choices', 'bool', 'integer', 'float',
                         'zip', 'date', 'pounds', 'year', 'month', 'day',
-                        'feet-inches', 'blood-pressure', 'choice-ref',
+                        'feet-inches', 'blood-pressure', 'choice-ref', 'scale',
```
You do not change files in the migration/models directory. You need to insert scale in your migration script.

au2 commented on migration/migrations/20180206155614-scale-question-type.js:
```diff
@@ -0,0 +1,11 @@
+        defaultValue: null
+    });
+  },
+};
```
Run eslint for this file. And insert 'scale' to question types here.

au2 commented:
We also need to update patch for this. Take a look at the question_patch tests and I think you can add new test for scale limits and go from there for implementation.

au2 commented:
Also please merge develop. Looks like my merges introduced conflicts with your development.

orndorffgrant commented:
Will do. Thank you for the help! I'll reach out if I have questions

orndorffgrant commented:
I think I got everything you mentioned @au2 
Specifically make sure I did the migration correctly, and that the patch tests are sufficient.

au2 commented on models/dao/question.dao.js:
```diff
@@ -623,6 +657,16 @@ module.exports = class QuestionDAO extends Translatable {
+                if (scaleLimits && question.scaleLimits && question.type === 'scale') {
+                    if (!_.isEqual(question.scaleLimits, scaleLimits)) {
+                        const combined = _.cloneDeep(question.scaleLimits);
+                        Object.assign(combined, scaleLimits);
```
I not sure what you are trying to accomplish here. Can't you just do parameterJsonToString({ scaleLimits }) without doing any clones or assigns?

orndorffgrant replied:
We don't allow nulls for min or max in patches? In that case, then yes this is unnecessary (as well as similar merging in the test patching).
That does limit the capabilities though: if a question had only a min on creation, then we wanted to update the min, we'd have to give it a max as well. Is that okay?

au2 commented:
Sorry for piecemeal comments but I think you will need to update surveyQuestionPatch object in swagger and add updates to survey_patch.model and survey_patch.integration tests. Take a look at those test, they are a but overengineered so let me know if you need any help understanding them. You need to add a scale type question to some of the surveys.

au2 commented:
Yes, it should be OK. Unfortunately Swagger 2.0 does not support null's for
fields. Open API 3.0 (aka Swagger 3.0) will support nulls but the library
we are using (swagger-tools) is not there yet.

On Wed, Feb 7, 2018 at 5:46 PM, Grant Orndorff <notifications@github.com>
wrote:

> *@orndorffgrant* commented on this pull request.
> ------------------------------
>
> In models/dao/question.dao.js
> <https://github.com/amida-tech/recruitment-registry/pull/249#discussion_r166781565>
> :
>
> > @@ -623,6 +657,16 @@ module.exports = class QuestionDAO extends Translatable {
>                      }
>                      return null;
>                  });
> +                const { scaleLimits } = patch;
> +                if (scaleLimits && question.scaleLimits && question.type === 'scale') {
> +                    if (!_.isEqual(question.scaleLimits, scaleLimits)) {
> +                        const combined = _.cloneDeep(question.scaleLimits);
> +                        Object.assign(combined, scaleLimits);
>
> We don't allow nulls for min or max in patches? In that case, then yes
> this is unnecessary (as well as similar merging in the test patching).
> That does limit the capabilities though: if a question had only a min on
> creation, then we wanted to update the min, we'd have to give it a max as
> well. Is that okay?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/amida-tech/recruitment-registry/pull/249#discussion_r166781565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF1-EUjPcVq02sXZZISTC0XodB_awsBLks5tSifZgaJpZM4R2F_Z>
> .
>

orndorffgrant commented:
@au2 I updated the surveyQuestionPatch object in swagger. 

It looks like the survey_patch tests generate questions using question-generator.js which generates scale type questions every X questions, so scale questions are being included in the test surveys. I verified that scale questions are present via console.log and grep. Do I need to do anything else?

au2 commented:
Was it failing before you made changes to surveyQuestionPatch?

On Fri, Feb 9, 2018 at 3:33 PM, Grant Orndorff <notifications@github.com>
wrote:

> @au2 <https://github.com/au2> I updated the surveyQuestionPatch object in
> swagger.
>
> It looks like the survey_patch tests generate questions using
> question-generator.js which generates scale type questions every X
> questions, so scale questions are being included in the test surveys. I
> verified that scale questions are present via console.log and grep. Do I
> need to do anything else?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/amida-tech/recruitment-registry/pull/249#issuecomment-364557265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF1-Ef435bRHAsqXcd3uv6ICm_RFaZUsks5tTKt-gaJpZM4R2F_Z>
> .
>

orndorffgrant commented:
Hmmm I just removed scaleLimits from surveyQuestionPatch and the tests still passed. What does that mean?

au2 commented:
That means you are not patching any of the scale questions in the test. You
need to patch at least have one case where you patch a scale type questions
(change min, max or both).

On Fri, Feb 9, 2018 at 3:56 PM, Grant Orndorff <notifications@github.com>
wrote:

> Hmmm I just removed scaleLimits from surveyQuestionPatch and the tests
> still passed. What does that mean?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/amida-tech/recruitment-registry/pull/249#issuecomment-364563924>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF1-EZF_uN1ULq9PvewQnp1oah3idTefks5tTLDigaJpZM4R2F_Z>
> .
>

orndorffgrant commented:
@au2 last commit should do it. Check additions to test/fixtures/conditional-session/patch/patch-setup.js and test/util/generator/survey-patch-generator.js

au2 commented:
1) Please run grunt or grunt eslint. Looks like you did not run eslint.
2) I am not sure how you found the scale question. If it is by trial and error I would update setup  to make a question of type scale explicitly.

orndorffgrant commented:
Thank you! I was unaware of to how to explicitly insert a question but you pointed me in the right direction. 

I think this is finally good to go @au2
